### DISCUSSION TITLE
Implementation of bot in this repository

### DISCUSSION THREAD

**Discussion by dhruvmanila** [TheAlgorithms/Python#8574] [Team Posts]
I'm currently working on implementing a bot: https://github.com/dhruvmanila/algobot
Currently, it can only greet the user who installed the app but I will add the functionality mentioned in [this comment](https://github.com/TheAlgorithms/Python/issues/3609#issuecomment-715052957) soon. The bot uses asynchronous programming in Python, GitHub REST API to make the calls and Heroku for hosting.

If anyone got any further ideas to be implemented or they want to help, please do tell.

P.S. Can anyone suggest a name and display photo for the bot?

**Comment by dhruvmanila** [comment]
Progress:
https://github.com/dhruvmanila/testing/pull/10

Completed the feature where the bot adds and removes the label when any PR fails or succeeds. Next thing to implement is to write proper tests for it.

**Comment by dhruvmanila** [comment]
So, the bot is now in a good condition (100% coverage on tests) to be installed, you can check it out at the above link.

@cclauss Do we really want to directly close the PR if it doesn't contain any tests or type hints? Also, this will only work if it contains zero tests and zero type hints.

**Comment by poyea** [comment]
It's also good for the bot to display those metrics in comments. If the none of the files changed end with `.py`, that PR MUST be closed. Not sure if it's easy to do for the checking.

**Comment by cclauss** [comment]
If a contributor suggests a change to one of our .md files or .yml files, I do not think we want to autoclose that PR.

**Comment by poyea** [comment]
@cclauss yeah I might state that wrongly: sometimes there're PRs with files without any file extension. We can deal with that case

**Comment by cclauss** [comment]
I think there should be some subtlety in how the bot behaves...  If the repo has > 50 open PRs then the bot should be more aggressive about autoclosing non-compliant PRs.  If there are few PRs then the bot should add labels to PRs and maintainers should guide the contributors to make appropriate corrections before the stalebot jumps it.  Is it possible for the bot to see how many open PRs the repo has?

Should we shorten the waiting times on stalebot?

**Comment by dhruvmanila** [comment]
This idea can be implemented: auto close the PR with files without any file extension or better to have a list of accepted file extension: `.py`, `.yml` or `.yaml`, `.md`, `.txt`

> Is it possible for the bot to see how many open PRs the repo has?

Kind of. GitHub REST API (v3) considers all pull requests as issues but not all issues are considered pull requests, so the number we get is the combination of open issues and pull requests. Maybe the GraphQL API (v4) has the ability to give us the exact information, I will have to look into that.

My question was whether to close the PR or label it appropriately when there are no type hints or any kind of tests.

**Comment by cclauss** [comment]
> whether to close the PR or label it appropriately

The bot should always label the PR.  My vote would be that if the repo has > 50 open PRs, the bot should also autoclose PR.

**Comment by dhruvmanila** [comment]
Another thing is that I'm changing the way I commit in the bot repository. I will be making different branches for related features and master will only be used for deployment which is done automatically after all the tests pass. All the constants related to the bot will be in either the `utils.py` file or `const.py` file like `FAILURE_LABEL`, `ACCEPTED_PR_FILE_EXTENSION`, `MIN_PR_FOR_AGGRESSIVE_BOT` and so on.

Some ideas when a pull request is opened:
- Check whether the description is empty or not, if it is then close the PR.
- Check whether the user has an open PR and if so then make an appropriate comment and close the PR. `MAX_PR_PER_USER=?`
- Check for type hints and tests in the submitted files and label it appropriately (close it when there are more number of PRs). I can get the exact number of open PRs via the search API endpoint. 
- Check for file extension and close the PR where the file extension is not included in the list. `.github/` directory will be ignored.
- After all the checks are completed, label the PR if the tests fail and comment on how to fix the errors.

Also, whatever ideas anyone get just dump it here. I will make it work if it is possible :)

**Comment by cclauss** [comment]
Check of no boxes are checked in the default commit message.

**Comment by dhruvmanila** [comment]
Yes, that is a good idea. But should we close the PR or label/comment it appropriately?

**Comment by cclauss** [comment]
An unmodified default commit message == an empty commit message in my book.  I would delete if we have a lot of open PRs.

**Comment by dhruvmanila** [comment]
Alright, got it

**Comment by dhruvmanila** [comment]
What are the criteria for marking a pull request as `invalid`?
- Empty description
- No checkbox ticked
- Invalid file format (Extensionless files)

Any other?

**Comment by dhruvmanila** [comment]
Woah! Take a look at this: https://github.com/dhruvmanila/testing/pull/15
This is basically what the bot will do:
- Create a PR report about tests, type hints, descriptive parameter names and function names and share it with the user if there are any errors.
- The report will be commented only when the PR is opened and not for the next commits.
- The labels will be added and removed accordingly.

Please open a PR to make the changes for the comments the bot will be posting. I just made the basic ones to test out the bot:
- Labels name: https://github.com/dhruvmanila/algobot/blob/event/pull_request/algobot/constants.py
- Comments: https://github.com/dhruvmanila/algobot/blob/event/pull_request/algobot/pull_requests.py

**Comment by dhruvmanila** [comment]
PR report will list out the node path to the specific class/function/parameters in this order:
`{filename}:{class_name}:{function_name}:{parameter_name}`

`class_name` will only be present if the function is inside a class (basically, if it is a method)

**Comment by cclauss** [comment]
!!!!

**Comment by dhruvmanila** [comment]
So, how's the bot up to now? :-)

One more thing which will be updated is that for a `class`, the `self` parameter should be skipped and private methods do not require `doctest` as private methods are mostly called from the public methods.

I will also implement the auto closing feature for issues whose description is empty. Then we won't need the workflow file to auto close empty issues.

**Comment by cclauss** [comment]
I do not see a downside to requiring tests on private methods.

**Comment by dhruvmanila** [comment]
Noted ✅

**Comment by dhruvmanila** [comment]
Also, if anyone can suggest a better name and an awesome display photo? 😅 😄

**Comment by cclauss** [comment]
https://en.wikipedia.org/wiki/Buzz_cut (all one word: `buzzcut`).  You should release on PyPI to reserve the name.

**Comment by dhruvmanila** [comment]
`buzzcut` as a bot name? Can you tell me why lol?

Also, the bot will be on the GitHub marketplace so the name will be reserved there instead. There’s no point in releasing a bot on pypi, atleast not now.

**Comment by dhruvmanila** [comment]
Hi @cclauss or anyone else, can you open a PR in https://github.com/dhruvmanila/testing ? I want to check one thing but that is not possible if I open the PR. Also, keep the description empty so that the PR gets closed automatically.

**Comment by cclauss** [comment]
A buzzcut is a radical removal of stuff that sticks out or gets in your way.  It implies a merciless action to get things down to the bare essentials.  It is a bit overstated but it is a short, memorable name nicely describes what your bot does.

Many project maintainers faced the same Hacktoberfest onslaught (30 PRs —> 530) that we did.  Your README should explain that you are an open source maintainer and that buzzcut allowed you to bring your project back to a health state after the Hacktoberfest hangover. 🤕 

Buzzcut puts a moat in place to help and your team of co-maintainers be ready for the onslaught of Hacktoberfest 2021.

**Comment by dhruvmanila** [comment]
Also, what about the dunder methods? I don't think that should have `doctest`. What's your take on that?

No point in having `doctest` for `__init__`

**Comment by cclauss** [comment]
I am OK with no doctest for `__init__()` but I believe that other dunder methods should have doctests as they can be implemented incorrectly.

**Comment by dhruvmanila** [comment]
Agreed. Implemented the ability to ignore `__init__` for `doctest`

**Comment by dhruvmanila** [comment]
The name `buzzcut` is already in use as a GitHub app. Our bot is basically a GitHub app so we need a unique name that is available there.

**Comment by dhruvmanila** [comment]
@cclauss The bot is ready except for the name and display photo. Do we ask the community for it?

**Comment by cclauss** [comment]
Prereview? MaintainersHelper? SanityCheck? Let’s get it in production even if we just call it AlgoBot.

**Comment by dhruvmanila** [comment]
I am sorry I do not understand what do you mean. Can you please explain in brief?

**Comment by dhruvmanila** [comment]
### Ideas:

- All checks will be ignored if the pull request is in draft mode. When the user makes the PR `ready for review`, the checks will start.
- Add the label for `requested changes` when a review has been submitted with requested changes option. Remove the label when the PR has been approved. For removing the label this logic doesn't sound right so I'm open to discussion.
- Add the label for `ready for review` to indicate that the PR is ready for review only after all tests are passing and there are no errors detected by the parser. This logic is a bit difficult to achieve but I will try my best. This can be extremely useful for maintainers so that they can spend their time on quality PRs.
- Cleanup pull request after it is closed but only if it is merged.
  - Delete the PR branch only if it originated from this repository
  - Remove all the labels?

**Comment by cclauss** [comment]
Nice!  I like this approach (draft, not close) better!!

When the PR is first opened if it fails some test, can the bot put that PR in draft mode instead of closing it?

The bot should only remove labels that are in its vocabulary so if a repo has other labels, the repo does not mess with them.

**Comment by dhruvmanila** [comment]
Yes it can. Also the bot is not going to close the PR if the checks fails, just going to add the label and I’m thinking of adding a comment with links to the logs. 

The bot will close the PR only if:
- A PR has no description 
- The user did not go through the checklist
- User has multiple PRs opened

> The bot should only remove labels that are in its vocabulary so if a repo has other labels, the repo does not mess with them.

The bot only does which it is programmed to do and nothing more. I have written in brief what it will do in the repository README. You can take a look at that.

**Comment by dhruvmanila** [comment]
Can we have a discussion at a time pre-decided? The long silence is very irritating lol.

**Comment by dhruvmanila** [comment]
@cclauss Can I have an update for the bot?

> Prereview? MaintainersHelper? SanityCheck? Let’s get it in production even if we just call it AlgoBot.

What do you mean by this? Also, the bot name is `algorithms-bot`, even `algobot` was not available. I still have to change the directory name. Want to keep the name as `gentleman`?

> - A PR has no description
> - The user did not go through the checklist
> - User has multiple PRs opened

The bot will close the PR only for the above three conditions. 

When the tests are failing, the bot will add/remove the label accordingly. Also, if there are any missing requirements such as no doctest or type hints are missing or function/parameter names are of length 1, the bot will make a report and comment it and label the PR accordingly, but only when the PR is opened or made ready for review. We don't want to clutter the PR with bot comments.

**Comment by dhruvmanila** [comment]
Got one more name: `algorithms-keeper`

**Comment by cclauss** [comment]
Can we do a Skype/Zoom/FaceTime/other call?

**Comment by dhruvmanila** [comment]
Does Google meet work for you? At 16:00 UTC?

**Comment by cclauss** [comment]
https://meet.google.com/smn-qajp-rkg

**Comment by dhruvmanila** [comment]
Hey, @cclauss you there?

**Comment by dhruvmanila** [comment]
With the help of @cclauss I was able to implement the feature where the bot will add a label to indicate the PR is ready to be reviewed by a maintainer (`Status: awaiting reviews`) if all the checks are passing and the PR satisfies all the requirements. It will be removed once the PR is reviewed.

Code: https://github.com/dhruvmanila/algorithms-keeper/blob/master/algorithms_keeper/pull_requests.py#L268

**Comment by dhruvmanila** [comment]
It seems that the `awaiting reviews` label does not need to be removed if there are missing requirements or the checks are failing. If there are any of the labels indicating that some requirements are missing or checks are failing then it is self-explanatory that the PR might not be ready to be reviewed at this moment.

This is my proposed workflow for labeling the status of the pull request:
- PR opened -> Add `awaiting reviews` label
- Reviewed and changes requested -> Add `awaiting changes` label
- Approved -> Remove all `awaiting ...` labels
- ***New commits after being approved -> Add `awaiting reviews` label ??***

~This is how the CPython repository workflow looks like (I have excluded the `awaiting core review` label as that is not useful for our purposes):~
~- PR opened -> Add `awaiting reviews` label~
~- Reviewed and changes requested -> Add `awaiting changes` label~
~- Changes made -> Add `awaiting change review` label~
~- Changes requested again -> Add `awaiting changes` label~
~- Approved -> Add `awaiting merge` label~

~Should we adopt this style of workflow or keep it simple with the previous workflow?
Or adopt only the `awaiting change review` part of the workflow as that seems to be useful for our purposes.~

Ignore this, we don't need this type of workflow.

Another thing I'm thinking is to let the bot comment regarding the missing requirements in every check as just the label part won't be enough for the user to understand. The later comments will include only the `content` part and not the explanation and relevant links section.

**Comment by dhruvmanila** [comment]
I have got another idea: instead of commenting, we will make the bot start a check run and then that will initiate the parser in the code and the parser will go through all the files for missing requirements. After that, the bot will annotate the missing requirements right in the code with just one sentence like `please provide type hints for 'num' parameter` and so on.

**Comment by cclauss** [comment]
On the labels... keep the flow simple.  `awaiting review` should be a Boolean for tests have passed.

On the comments... make them as helpful and foolproof as possible.  The more explicit the comments can be about the requested change, the better.

I would like us to réengage with mypy even if it is for just one directory to start with.

**Comment by dhruvmanila** [comment]
This is my proposed changes for the label:
- PR open -> Apply `awaiting reviews` label
- Changes requested -> Apply `awaiting changes` label
- New changes pushed -> Apply `awaiting change review` label
- Changes requested again -> Apply `awaiting changes` label
- Approved -> Remove all `awaiting ...` label

The bot shouldn't remove the `awaiting reviews` label if the tests are failing or some `require ...` labels are present.

As for the comments, how about this?
> Another thing I'm thinking is to let the bot comment regarding the missing requirements in every check as just the label part won't be enough for the user to understand. The later comments will include only the content part and not the explanation and relevant links section.

**Comment by cclauss** [comment]
Please make the choice that you feel most comfortable with.  For me, I want one label that I can click on to see all (and only) PRs that are ready to be reviewed.  That label tells me that all tests have passed and that changes have been made to address any previous review comments.  Having multiple flavors of `awaiting review` makes the search more complicated for the maintainer.

**Comment by dhruvmanila** [comment]
>  For me, I want one label that I can click on to see all (and only) PRs that are ready to be reviewed.

Yes, then it makes sense to remove the `awaiting label` if the PR does not meet our requirements. Also, sometimes a user would need help in resolving the issues, I will include a line like `Tag a maintainer to ask for help`

`awaiting review` => The PR is never reviewed
`awaiting change review` => The PR was reviewed, changes were requested, user made the changes, needs to be reviewed again.

**Comment by dhruvmanila** [comment]
[PR stages image](https://imgur.com/U7VQHYs)

Take a look at this.

**Comment by dhruvmanila** [comment]
> On the labels... keep the flow simple. awaiting review should be a Boolean for tests have passed.

Currently, `awaiting review` is also removed if any of the `require ...` labels are added.

**Comment by dhruvmanila** [comment]
Edit: A much simpler workflow then previously suggested: https://imgur.com/74IAXG5
@cclauss Please take a look at this so that I can implement this right away.

**Comment by cclauss** [comment]
The only thing missing from the flow is the change of status when automated tests fail.

New PR --> Awaiting Review --> automated tests fail --> Awaiting Changes

**Comment by dhruvmanila** [comment]
This is the workflow around `tests are failing` label:

New PR -> `awaiting review` label -> Tests are failing -> Remove the `awaiting review` label -> Tests are passing -> Add back the `awaiting review` label.

`awaiting changes` label is added only when a maintainer requests a change in the PR.

There is `tests are failing` label which should suggest users to take a look at why they are failing and make the necessary changes. I am thinking about putting the CI logs where there are failures as a comment.

**Comment by dhruvmanila** [comment]
@cclauss @poyea If possible, please re-approve the PR if there were commits made after being approved.

It makes sense that if the PR is approved at the current state and then there are any changes added the PR needs to be reviewed and approved again. That's the way the bot understands :)

As a safety net I will remove all `awaiting...` labels after the PR is merged.

**Comment by dhruvmanila** [comment]
Hi there!

There is one other way the bot could represent the missing requirement information in the pull request:
1. Through Checks API by annotating the code directly. Take a look at [this video](https://docs.github.com/en/free-pro-team@latest/developers/apps/creating-ci-tests-with-the-checks-api) (scroll down a bit and the first image is the video):

![image](https://docs.github.com/assets/images/github-apps/github_apps_checks_annotations.png)

2. Make review comments just like how we do it in the pull request: https://github.com/theforeman/foreman/pull/8136 (Scroll down a bit and you will notice it by `houndci` which is a GitHub app). The status for checks can be seen at the bottom of the pull request where all the checks resides. We can update the status to `failure`, `success` accordingly.

**Comment by dhruvmanila** [comment]
I'm going with option number 2: https://github.com/dhruvmanila/testing/pull/34#pullrequestreview-541782450

Check run annotation will only be visible in the files section and the checks section which is one click away for the author. We want to make it easy for them to see what is missing in their submission and review comments seems to be the best way. The code is visible, where is it missing is visible, the file name will be provided in the comment.

On a side note, I was rewriting the parser and there's going to be a [`Visitor` class](https://github.com/dhruvmanila/algorithms-keeper/blob/parser-rewrite/algorithms_keeper/parser_rewrite.py#L262) which can visit any node in the tree (source code), and thus, we can perform checks for any rules we define. So, if anyone has other ideas regarding any of the node such as `lambda`, `comprehension`, `variables`, `assignments`, `operators`, etc., feel free to dump it here.
